src/lib/util/corestr.cpp

```diff
@@ -206,12 +208,49 @@ std::string &strtrimrightspace(std::string& str)
 	return internal_strtrimspace(str, true);
 }
 
-std::string &strmakeupper(std::string& str)
+static std::string &strtransform_uchar(std::string& str, char32_t(*callback)(char32_t ch))
 {
-	std::transform(str.begin(), str.end(), str.begin(), ::toupper);
+	size_t i = 0;
+	while(i < str.size())
+	{
+		size_t advance;
+		char32_t old_character;
+		int rc = uchar_from_utf8(&old_character, &str[i], str.size() - i);
+
+		if (rc > 0)
+		{
+			char32_t new_character = callback(old_character);
+			if (old_character != new_character)
+			{
+				// if the character has changed, replace it
+				std::string new_character_string = utf8_from_uchar(new_character);
```

Comment:
This is horribly inefficient - you're doing moves all over the place.  You're better off operating on ranges.  Start with an empty transformed string, when you get a character that needs transforming, copy the entire range from start to the changing character, then append the changed character, etc.  If you want to get fancy, you can also copy ranges only at points where the size of the UTF-8 sequence changes.